src/mesh/invert3x3.hxx

```diff
@@ -29,42 +29,38 @@
 #pragma once
 
 #include <bout/utils.hxx>
+#include <optional>
 
 /// Explicit inversion of a 3x3 matrix \p a
 ///
-/// The input \p small determines how small the determinant must be for
-/// us to throw due to the matrix being singular (ill conditioned);
-/// If small is less than zero then instead of throwing we return false.
-/// This is ugly but can be used to support some use cases.
-template <typename T>
-bool invert3x3(Matrix<T>& a, T small = 1.0e-15) {
+/// If the matrix is singular (ill conditioned), the determinant is
+/// return. Otherwise, an empty `std::optional` is return
+namespace bout {
+inline std::optional<BoutReal> invert3x3(Matrix<BoutReal>& a) {
   TRACE("invert3x3");
 
   // Calculate the first co-factors
-  T A = a(1, 1) * a(2, 2) - a(1, 2) * a(2, 1);
-  T B = a(1, 2) * a(2, 0) - a(1, 0) * a(2, 2);
-  T C = a(1, 0) * a(2, 1) - a(1, 1) * a(2, 0);
+  BoutReal A = a(1, 1) * a(2, 2) - a(1, 2) * a(2, 1);
+  BoutReal B = a(1, 2) * a(2, 0) - a(1, 0) * a(2, 2);
+  BoutReal C = a(1, 0) * a(2, 1) - a(1, 1) * a(2, 0);
 
   // Calculate the determinant
-  T det = a(0, 0) * A + a(0, 1) * B + a(0, 2) * C;
-
+  const BoutReal det = a(0, 0) * A + a(0, 1) * B + a(0, 2) * C;
+  constexpr BoutReal small = 1.0e-15;
   if (std::abs(det) < std::abs(small)) {
-    if (small >= 0) {
-      throw BoutException("Determinant of matrix < {:e} --> Poorly conditioned", small);
-    }
-    return false;
+    return std::optional<BoutReal>{det};
```

Comment:
Is the explicit constructor really needed? Could this not be:
```suggestion
    return det;
```

Reply:
Not sure, will check

Reply:
:innocent: Did you check?
I am also fine with it being explicit, so this can be merged ...